File: src/C/GNC.c
#include <stdbool.h>
#include <math.h>

struct input;
struct input
{
    double r[3];
    double v[3];
    double m_prop;
};

struct commands;
struct commands
{
    bool NOOP;
};

struct parameters;
struct parameters
{
    double mu;
};

struct output;
struct output
{
    bool thr_on;
    double thr_dir[3];
    double specific_energy;
};

struct output step(const struct input *i, struct commands *c, const struct parameters *p)
{
    bool thr_on;
    double thr_dir[3];
    double specific_energy;

    double R2 = 0.0;
    double V2 = 0.0;
    for (int j = 0; j < 3; j++)
    {
        R2 += i->r[j] * i->r[j];
        V2 += i->v[j] * i->v[j];
    }

    specific_energy = V2 / 2 - p->mu / sqrt(R2);

    thr_on = (i->m_prop > 0) && (specific_energy < 0);

    if (thr_on && V2 > 0)
    {
        double V = sqrt(V2);

        for (int j = 0; j < 3; j++)
        {
            thr_dir[j] = i->v[j] / V;
        }
    }
    else
    {
        thr_dir[0] = 0;
        thr_dir[1] = 0;
        thr_dir[2] = 0;
    }

    struct output o = {
        .thr_on = thr_on,
        .thr_dir = *thr_dir,
        .specific_energy = specific_energy,
    };

    return o;
}

File: src/GNC.jl
module GNC

using ..Types

using LinearAlgebra

using DynamicalSystems, Unitful

using BenchmarkTools


struct input
    r::Position     # Current position
    v::Velocity     # Current velocity
    m_prop::Mass    # Current propellant mass
end

struct commands
end

struct parameters
    μE::GravitationalParameter
end

struct output
    thr_on::Bool
    thr_dir::SVector{3,Float64}
    specific_energy::SpecificEnergy
end

function step(i::input, c::commands, p::parameters)::Tuple{output,commands}

    ϵ0 = ϵ(i.r, i.v, p.μE)

    thr_on = (i.m_prop > 0.0u"kg") && (ϵ0 < 0.0u"m^2/s^2")

    if thr_on
        thr_direction = normalize(ustrip.(i.v))
    else
        thr_direction = SVector(0.0, 0.0, 0.0)
    end

    return (
        output(thr_on, thr_direction, ϵ0),
        commands()
    )

end

"""
    ϵ(r::Position, v::Velocity, μ::GravitationalParameter)

Orbital specific energy (m^2/s^2)
"""
function ϵ(r::Position, v::Velocity, μ::GravitationalParameter)

    R = norm(r)
    V = norm(v)

    return upreferred(V^2 / 2 - μ / R)
end


function test_run()

    RE = upreferred(6371.009u"km")
    μE = upreferred(398600.4418u"km^3/s^2")
    R0 = upreferred(RE + 400u"km")
    V0 = upreferred(sqrt(μE / R0))

    r0 = Position(R0, 0.0u"m", 0.0u"m")
    v0 = Velocity(0.0u"m/s", V0, 0.0u"m/s")
    mp = 100.0u"kg"

    gnc_i = input(r0, v0, mp)
    gnc_c = commands()
    gnc_p = parameters(μE)

    gnc_o, gnc_c = step(gnc_i, gnc_c, gnc_p)

    return (gnc_i, gnc_c, gnc_p, gnc_o)

end

end